llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,386 @@
+
+template <class Container> std::string Join(const Container &C, char sep) {
+  std::string S;
+  auto it = C.begin();
+  for (size_t i = 0; i != C.size(); ++i, ++it) {
+    if (i > 0)
+      S += ", ";
+
+    S += std::to_string(*it);
+  }
+
+  return S;
+}
+
+template <class Container>
+void CheckDeclaredAspects(LLVMContext &C, const Function *F,
+                          const Container &Aspects) {
+  MDNode *MDN = F->getMetadata("intel_declared_aspects");
+  if (!MDN)
+    return;
+
+  AspectsSetTy DeclaredAspects;
+  for (unsigned i = 0; i != MDN->getNumOperands(); ++i) {
+    const ConstantAsMetadata *CM =
+        dyn_cast<const ConstantAsMetadata>(MDN->getOperand(i));
+    DeclaredAspects.insert(
+        dyn_cast<const ConstantInt>(CM->getValue())->getSExtValue());
+  }
+
+  SmallVector<int> MissedAspects;
+  for (int aspect : Aspects) {
+    if (DeclaredAspects.count(aspect) == 0) {
+      MissedAspects.push_back(aspect);
+    }
+  }
+
+  if (!MissedAspects.empty()) {
+    std::string AspectsStr = Join(MissedAspects, ',');
+    // TODO: demangle function name and aspect's IDs?
+    std::string Msg = formatv(
+        "for function \"{0}\" there is the list of missed aspects: [{1}]",
+        F->getName(), AspectsStr);
+
+    emitWarning(C, Msg);
+  }
+}
+
+using FunctionToAspectsMapTy = DenseMap<Function *, SmallSet<int, 4>>;
+using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
+
+void CreateAspectMetadataForFunctions(FunctionToAspectsMapTy &Map) {
+  for (auto &it : Map) {
+    Function *F = it.first;
+    auto &Aspects = it.second;
+    if (Aspects.empty())
+      continue;
+
+    LLVMContext &C = F->getContext();
+    std::vector<Metadata *> AspectsMetadata;
+    AspectsMetadata.reserve(Aspects.size());
+    for (int aspect : Aspects) {
+      AspectsMetadata.push_back(ConstantAsMetadata::get(
+          ConstantInt::getSigned(Type::getInt32Ty(C), aspect)));
+    }
+
+    MDNode *MDN = MDNode::get(C, AspectsMetadata);
+    F->setMetadata("intel_used_aspects", MDN);
+  }
+}
+
+void CheckAllDeclaredAspects(FunctionToAspectsMapTy &Map) {
+  for (auto &it : Map) {
+    Function *F = it.first;
+    auto &Aspects = it.second;
+    if (Aspects.empty())
+      continue;
+
+    LLVMContext &C = F->getContext();
+    CheckDeclaredAspects(C, F, Aspects);
+  }
+}
+
+void PropagateAspects(Function *F, CallGraphTy &CG,
+                      FunctionToAspectsMapTy &AspectsMap,
+                      SmallPtrSet<Function *, 16> &Visited) {
+  SmallSet<int, 8> LocalAspects;
+  for (Function *Callee : CG[F]) {
+    if (Visited.contains(Callee)) {
+      Visited.insert(Callee);
+      PropagateAspects(Callee, CG, AspectsMap, Visited);
+    }
+
+    auto &CalleeAspects = AspectsMap[Callee];
+    LocalAspects.insert(CalleeAspects.begin(), CalleeAspects.end());
+  }
+
+  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+FunctionToAspectsMapTy
+GetFunctionsAspects(Module &M, TypeToAspectsMapTy &TypesWithAspects) {
```

Comment:
```suggestion
GetFunctionsAspects(const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
```

Reply:
It can't be done easily because it would lead to change `Function &` to `const Function &` in loop. And it would lead to change in `FunctionToAspectsMapTy` type (Key would be required to be `const Function *`).